### DISCUSSION TITLE
Impove useAsyncData

### DISCUSSION THREAD

**Discussion by Mini-ghost** [nuxt/nuxt#17765] [Ideas]
In current, we can use `useAsyncData` to fetch data at server side or client side.  this will cache and  use key to find data on payload to hydration. even can use options that `pick` and `transform` to get we need data.

In some case, we can fetch same api with same key between different component. But if we has two `useAsyncData` with same key on different components, and one of be called in server side, the other one be called on client side, or we need data with different structure (with pick and transform options), that will got some problems.

In PR #3985, resolve useAsyncData invoke between server and client issues. but the issue of data structure still exist.

In current, nuxt will cache data that parsed by `pick` and `transform`, this causes the following two components to work unexpectedly:

**A.vue**
```vue
<template>
  <div>
    {{ text }}
  </div>
</template>

<script setup lang="ts">
import axios from 'axios'
const fetchFn = () => axios('https://pokeapi.co/api/v2/generation/1/').then(res => res.data)
const { data: text } = useAsyncData('HELLO_WORLD', () => fetchFn(), {
  pick: ['pokemon_species']
})
</script>
```

**B.vue**
```vue
<template>
  <div>
    {{ text }}
  </div>
</template>

<script setup lang="ts">
import axios from 'axios'
const fetchFn = () => axios('https://pokeapi.co/api/v2/generation/1/').then(res => res.data)
const { data: text } = useAsyncData('HELLO_WORLD', () => fetchFn())
</script>
```

Both A.vue and B.vue component will get same data even with different `pick` option.

Maybe we could manage all fetch original result. Subsequent invoke useAsyncData with same key, deep clone from cached original result and return new data after resolve by pick or transfrom. And we need a method to remove result cached, when cached data is no longer needed.

Maybe we could manage all fetch original result, then when invoke useAsyncData with same key, deep clone from cached original result and return the new data be resolved by `pick` or `transform`. And we need a method to remove cached result when it is no longer needed.

Week ago, I tried to resolve this issues in my fork repository (https://github.com/Mini-ghost/framework/tree/feat/use-async-data-improve). If possible,  I could create a PR to improve it.

Sorry with poor english, if there is something misunderstanding, I am so glad to see the comment to discuss.

**Comment by danielroe** [comment]
useAsyncData should not be called with the same key if it is meant to resolve to different data. However, you always create a custom composable that modifies the key based on the options, like `useFetch` does. It's also worth considering for Nuxt as well, as it may solve hard-to-diagnose issues.

The use-case for `pick` and `transform` is to avoid passing extra data in the payload, so caching the extra data would probably not be the best option.

**Reply by Mini-ghost** [reply]
If `useAsyncData` can help to management original result of fetch, maybe could reduce a lot of request. like pokeapi api, we need different part data from same result in many different components, we can cache original result when first fetch, when other component need other part of same result, we can avoid fetch again.

**Reply by Mini-ghost** [reply]
If want to avoid extra data, maybe could do like below:

```ts
import axios from 'axios'

// const fetchFn = () => axios('https://pokeapi.co/api/v2/generation/1/').then(res => res.data)
// const { data: text } = useAsyncData('HELLO_WORLD', () => fetchFn(), {
//   pick: ['pokemon_species']
// })

const fetchFn = () => axios('https://pokeapi.co/api/v2/generation/1/').then(res => res.data.pokemon_species)
const { data: text } = useAsyncData('POKEMON', () => fetchFn())
```
